src/object/object_primitive.c

```diff
@@ -13999,33 +14108,82 @@ mr_writeval_varnchar_internal (OR_BUF * buf, DB_VALUE * value, int align)
 	}
       else
 	{
-	  if (align == INT_ALIGNMENT)
+	  if (value->data.ch.medium.was_compressed == 0)
+	    {
+	      rc = pr_do_db_value_string_compression (value);
+	    }
+
+	  if (rc != NO_ERROR)
 	    {
-	      rc = or_packed_put_varchar (buf, str, src_size);
+	      return rc;
+	    }
+
+	  if (value->data.ch.medium.compressed_length == 0)
+	    {
+	      rc = pr_write_uncompressed_string_to_buffer (buf, str, src_size, align);
 	    }
 	  else
 	    {
-	      rc = or_put_varchar (buf, str, src_size);
+	      /* String has been prompted to compression before. */
+	      assert (value->data.ch.medium.was_compressed == 1);
+	      if (value->data.ch.medium.compressed_buf == NULL)
+		{
+		  assert (value->data.ch.medium.compressed_length == 0);
+		  string = value->data.ch.medium.buf;
+		}
+	      else
+		{
+		  assert (value->data.ch.medium.compressed_length != 0);
+		  string = value->data.ch.medium.compressed_buf;
+
+		}
+	      size = value->data.ch.medium.compressed_length;
+	      rc = pr_write_compressed_string_to_buffer (buf, string, size, src_size, align);
```

Comment:
Please check this call in context of first case : 
size = value->data.ch.medium.compressed_length = 0
I think we should write original/uncompressed buffer  (value->data.ch.medium.buf), with size = value->data.ch.medium.size